### PR TITLE
Improve diagnostic for variable assignment

### PR DESCRIPTION
**Pull request by singalhimanshu**
Now we report a missing equal error for the following code:

    let x new Array();

Fixes: #304

### CODE REVIEW

**Comment by singalhimanshu** [comment]
I forgot to add the docs. I will do that till tomorrow.

File: src/quick-lint-js/error.h [hunk 1]
```diff
@@ -1139,6 +1139,12 @@
       { source_code_span continue_statement; },                                \
       .error(QLJS_TRANSLATABLE("continue can only be used inside of a loop"),  \
              continue_statement))                                              \
+                                                                               \
+  QLJS_ERROR_TYPE(                                                             \
+      error_missing_equal_after_variable, "E151",                              \
+      { source_code_span expected_equal; },                                    \
+      .error(QLJS_TRANSLATABLE("missing '=' after variable"), expected_equal)) \
```
Comment (strager) [inline src/quick-lint-js/error.h:1146]: Nit: Sort alphabetically by `error_missing_equal_after_variable`.

File: src/quick-lint-js/parse.h [hunk 2]
```diff
@@ -3006,6 +3006,19 @@ class parser {
           break;
         }
 
+        case token_type::kw_new: {
+          const char8 *here = this->lexer_.end_of_previous_token();
+          this->error_reporter_->report(error_missing_equal_after_variable{
```
Comment (strager) [inline src/quick-lint-js/parse.h:3011]: Blocking: This change will reject the following legal code:
```js
let x
new Array()
```

We should only report an error if a semicolon cannot be inserted (i.e. if there is no newline before `new`).

File: src/quick-lint-js/parse.h [hunk 3]
```diff
@@ -3006,6 +3006,19 @@ class parser {
           break;
         }
 
+        case token_type::kw_new: {
```
Comment (strager) [inline src/quick-lint-js/parse.h:3009]: We should also report an error for other tokens, such as `class` and `typeof`.

File: src/quick-lint-js/parse.h [hunk 4]
```diff
@@ -3006,6 +3006,19 @@ class parser {
...
+          const char8 *here = this->lexer_.end_of_previous_token();
+          this->error_reporter_->report(error_missing_equal_after_variable{
+              .expected_equal = source_code_span(here, here),
+          });
+          expression *ast = this->parse_expression_remainder(
+              variable,
+              precedence{.commas = false, .in_operator = allow_in_operator});
+          this->lexer_.insert_semicolon();
```
Comment (strager) [inline src/quick-lint-js/parse.h:3017]: Hmm, inserting a semicolon after the expression unconditionally seems wrong. What about the following code?
```js
let a new Array(), b = a.length;
```
I think your change will cause an error at `,` (in addition to the expected error between `a` and `new`).

**Review by strager — CHANGES_REQUESTED** [review]

File: src/quick-lint-js/parse.h [hunk 5]
```diff
@@ -3026,6 +3026,23 @@ class parser {
...
+        case token_type::kw_new: {
+          bool has_leading_newline = this->peek().has_leading_newline;
+          if (!has_leading_newline) {
+            const char8 *here = this->lexer_.end_of_previous_token();
+            this->error_reporter_->report(error_missing_equal_after_variable{
+                .expected_equal = source_code_span(here, here),
+            });
+          }
+          this->visit_binding_element(variable, v, declaration_kind);
+          this->skip();
```
Comment (strager) [inline src/quick-lint-js/parse.h:3040]: Why are we skipping the `class` keyword? `class { potato }` is not the same as `{ potato }`.

File: test/test-parse-var.cpp [hunk 6]
```diff
@@ -498,6 +511,42 @@ TEST(test_parse, parse_invalid_let) {
                             compound_assignment_operator),  //
             declaring_token, offsets_matcher(&code, 0, u8"let"))));
   }
+
+  {
+    for (string8 kw : {u8"typeof", u8"class", u8"new"}) {
+      spy_visitor v;
+      padded_string code(u8"let x " + kw + u8" Array;");
```
Comment (strager) [inline test/test-parse-var.cpp:518]: What does `let x = class Array;` even mean? This test looks dubious.

File: test/test-parse-var.cpp [hunk 7]
```diff
@@ -498,6 +511,42 @@ TEST(test_parse, parse_invalid_let) {
...
+      parser p(&code, &v);
+      p.parse_and_visit_module(v);
+      EXPECT_THAT(v.visits, ElementsAre("visit_variable_declaration",  // x
+                                        "visit_variable_use",          // Array
```
Comment (strager) [inline test/test-parse-var.cpp:522]: Blocking: `typeof Array` should emit a `visit_variable_typeof_use`.

File: src/quick-lint-js/parse.h [hunk 8]
```diff
@@ -3026,6 +3026,23 @@ class parser {
...
+          this->parse_and_visit_expression(
+              v, precedence{.commas = false, .in_operator = allow_in_operator});
```
Comment (strager) [inline src/quick-lint-js/parse.h:3042]: Blocking: I think this breaks parsing of the following code:
```js
let x
class C {}
x = new C();
```
Please add a test and fix this. I think the problem is that you're parsing and visiting an expression even if `has_leading_newline` is true. If `has_leading_newline` is true, we should automatically insert a semicolon and end variable declaration statement parsing.

File: src/quick-lint-js/parse.h [hunk 9]
```diff
@@ -3026,6 +3026,23 @@ class parser {
               });
           goto initialize_variable;
 
+        case token_type::kw_typeof:
+        case token_type::kw_class:
+        case token_type::kw_new: {
```
Comment (strager) [inline src/quick-lint-js/parse.h:3031]: Nit: Sort cases alphabetically.
```suggestion
        case token_type::kw_class:
        case token_type::kw_new:
        case token_type::kw_typeof: {
```

**Review by strager — CHANGES_REQUESTED** [review]

File: src/quick-lint-js/parse.h (hunk 5 above)
Comment (singalhimanshu) [reply src/quick-lint-js/parse.h:3040]: Yes that was wrong I have fixed it now.

**Review by strager — APPROVED** [review]
Looks good to me! Let me know when I should merge.

**Comment by singalhimanshu** [comment]
> Looks good to me! Let me know when I should merge.

I guess we can merge it.

**Pull request merged**